mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/voice/VoiceInputCombinator.kt

```diff
@@ -20,36 +26,76 @@ object VoiceInputCombinator {
+              token().map { token ->
+                rulesSet
+                    .map { chessRule ->
+                      // Apply rule
+                      val res = chessRule.rule(token)
+                      res
+                    }
+                    .firstOrNull { res ->
+                      // Get the 1st value that have been transformed
+                      res != null
+                    }
+              })
+          .map {
+            when (it) {
+              "king" -> Rank.King
+              "queen" -> Rank.Queen
+              "rook" -> Rank.Rook
+              "bishop" -> Rank.Bishop
+              "knight" -> Rank.Knight
+              "pawn" -> Rank.Pawn
+              else -> null
+            }
+          }
+
+  val column =
+      convertToken(ChessSpeechEnglishDictionary.letters).filter { it in 'a'..'h' }.map {
+        it!! - 'a'
+      }
+
+  val row =
+      convertToken(ChessSpeechEnglishDictionary.numbers)
+          .filter { it in '0'..'9' }
+          .map {
+            // TODO : What happen id it is null ?
+            8 - (it!! - '0')
+          }
+          .filter { it in 0 until Board.Size }
+
+  val position =
       combine(
-          token("king").map { Rank.King },
-          token("queen").map { Rank.Queen },
-          token("rook").map { Rank.Rook },
-          token("bishop").map { Rank.Bishop },
-          token("knight").map { Rank.Knight },
-          token("pawn").map { Rank.Pawn })
+          CommonNotationCombinators.position,
+          CommonNotationCombinators.computePosition(column, row))
 
   /** A [Parser] which indicate the action between 2 position */
   private val actionSeparator = token("to")
 
   /** A [Parser] for a [Move] action. */
   private val move =
-      rank
-          .flatMap {
-            CommonNotationCombinators.position.flatMap { from ->
-              actionSeparator.flatMap {
-                CommonNotationCombinators.position.map { to -> Move(from, to) }
+      rank.flatMap {
+        if (it == null) {
+              failure()
+            } else {
+              position.flatMap { from ->
+                actionSeparator.flatMap { position.map { to -> Move(from, to) } }
               }
             }
-          }
-          .checkFinished()
+            .checkFinished()
+      }
 
   /** A [Parser] for a [Promote] action. */
   private val promote =
-      CommonNotationCombinators.position // No leading rank because only pawns may be promoted.
+      position // No leading rank because only pawns may be promoted.
           .flatMap { from ->
             actionSeparator.flatMap {
-              CommonNotationCombinators.position.flatMap { to ->
-                rank.map { rank -> Promote(from, to, rank) }
+              position.flatMap { to ->
+                // remove word that cannot be converted
+                rank.filter { it != null }.map { rank -> Promote(from, to, rank!!) }
```

Comment:
```suggestion
                rank.filterNotNull().map { rank -> Promote(from, to, rank!!) }
```

Reply:
Fix in https://github.com/epfl-SDP/android/pull/311/commits/2e5d3dd03e1f172cd12b388da540f44b50eb3984